refactor(apple): Migrate iOS/macOS clients to UniFFI

klochowicz opened the pull request:
Replace callback-based Adapter with event polling-based AdapterUniFfi

This change improves reliability by eliminating callback lifetime issues.

klochowicz committed:
refactor(apple): Migrate iOS/macOS clients to UniFFI

Replace callback-based Adapter with event polling-based AdapterUniFfi

This change improves reliability by eliminating callback lifetime issues.

jamilbk commented on scripts/build/uniffi-ios.sh:
```diff
@@ -0,0 +1,131 @@
+try_build "x86_64-apple-darwin" "macOS (x86_64)"
+try_build "aarch64-apple-darwin" "macOS (ARM64)"
+
+if [ ${#BUILT_TARGETS[@]} -eq 0 ]; then
```
I don't believe the iOS simulator supports Network Extensions correct? Would it make sense to exclude that target from our build scripts?

klochowicz replied:
Network Extensions won't run, but it might still be useful to start it for testing UI components etc.

jamilbk commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/Log.swift:
Curious why the changes to logging here?

thomaseizinger replied:
Seconding this, can we move these to a separate PR or are they required?

jamilbk commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -45,16 +61,24 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
     }
 
+    // Log version on actual tunnel start
+    let version =
```
I noticed the logic for getting the bundle version is duplicated - would it make sense to consolidate this to the SharedHelpers file?

klochowicz replied:
I'll leave this for a follow-up, ok? 

there's quite a few places where we also retrieve the version (UpdateNotification, Telemetry)  and it would be good to consolidate them all.

jamilbk reviewed (commented):
Looking good! Left some feedback after a first pass. Will want to test out this new build quite thoroughly I imagine.

thomaseizinger commented on rust/client-ffi/.cargo/config.toml:
```diff
@@ -0,0 +1,12 @@
+
+[build]
+# Use incremental compilation for faster rebuilds
+incremental = true
```
Missing newline.

klochowicz replied:
ah, I really need to set up something that does this automatically. I guess the pre-commit stuff should do the trick.

thomaseizinger replied:
Most editors have a setting for it :)

thomaseizinger commented on rust/client-ffi/src/platform/ios/make_writer.rs:
```diff
@@ -61,7 +67,7 @@ mod oslog {
                 match level {
                     tracing::Level::TRACE => ::oslog::Level::Debug,
-                    tracing::Level::DEBUG => ::oslog::Level::Info,
+                    tracing::Level::DEBUG => ::oslog::Level::Debug,
```
Why this change?

klochowicz replied:
temp hack, which I forgot to revert 🙈  - I was adding lots of debug logs on higher level and wanted to seet hem more easily.  thanks for spotting.

thomaseizinger commented on rust/client-ffi/src/platform/ios/make_writer.rs:
```diff
@@ -84,11 +90,11 @@ mod oslog {
     impl OsLog {
         pub fn with_level(&self, _: tracing::Level, _: &str) {
-            unimplemented!("Stub should never be called")
+            unimplemented!("oslog is only available on macOS/iOS")
```
Nit: We could move these changes to a separate patch-set to reduce the diff here.

thomaseizinger commented on rust/client-ffi/src/platform/ios/tun.rs:
```diff
@@ -17,14 +18,20 @@ pub struct Tun {
         set_non_blocking(fd)?;
         let name = name(fd)?;
 
+        tracing::info!("Creating TUN device from fd {} ({})", fd, name);
```
You are pretty liberal on the use of INFO logs, I've probably make all of these DEBUG.

thomaseizinger replied:
Also see https://github.com/firezone/corp/blob/main/architecture/madr/0012.md

klochowicz replied:
cheers, thanks for pointing me into the direction of ADRs!

thomaseizinger replied:
We don't use them very extensively but every now and then, I think of writing one when we make a new decision.

thomaseizinger commented on rust/client-ffi/src/platform/ios/tun.rs:
```diff
@@ -43,31 +50,61 @@ impl Tun {
                     "Failed to send to TUN device: {}"
-                )
+                );
+                tracing::error!("TUN send thread exited unexpectedly!");
```
We should probably move this to `tun_send` so it is present on all platforms.

klochowicz replied:
I've ripped this out from this PR to make it easier to review. Will post these changes separately.

thomaseizinger commented on rust/client-ffi/src/platform/apple/tun.rs:
Do we _need_ all of these changes? Would be nice to make the PR as minimal as possible.

klochowicz replied:
moved these out of this PR.

thomaseizinger commented on rust/client-ffi/src/platform/apple.rs:
This file should probably be called `apple.rs`?

thomaseizinger commented on rust/client-ffi/src/lib.rs:
```diff
@@ -114,6 +114,42 @@ impl Session {
+        clippy::too_many_arguments,
+        reason = "This is the API we want to expose over FFI."
+    )]
+    pub fn new_ios(
```
```suggestion
    pub fn new_apple(
```

thomaseizinger commented on scripts/build/uniffi-ios.sh:
```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env bash
+
+# Build script for UniFFI iOS bindings with XCFramework
+# This builds the Rust library for all iOS/macOS architectures and creates an XCFramework
```
This comment seems to be outdated because it still mentions macOS?

thomaseizinger commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Managers/SystemExtensionManager.swift:
These changes could also be moved to a separate PR I guess?

thomaseizinger commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Views/UpdateNotification.swift:
Are these just formatting changes?

klochowicz replied:
yes, it's weirdly going back and forth between them. just formatting. will revert it.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/ConnlibUniFfi.swift:
This file is completely empty. How does this do anything?

klochowicz replied:
this should have been removed.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/AdapterUniFfi.swift:
```diff
@@ -0,0 +1,569 @@
+}
+
+/// Adapter using UniFFI polling-based event handling
+class AdapterUniFfi: @unchecked Sendable {
```
I think we can just call this `Adapter`. The other one is deleted so we don't really need to differentiate.

klochowicz replied:
moved to 'Adapter.swift' and brought back the old name.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/AdapterUniFfi.swift:
```diff
@@ -0,0 +1,569 @@
+      var tunSetSuccessfully = false
+      for attempt in 1...Self.tunSetupMaxAttempts {
+        do {
+          try session.setTunFromSearch()
```
Can we just do this implicitly as part of `connect`? It seems odd that we do it asynchronously here. Alternatively, the FD should already be available as soon as the network extension is up so we could also do it in Swift and then call `setTun` directly after `connect`.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/AdapterUniFfi.swift:
```diff
@@ -0,0 +1,569 @@
+      )
+    } catch {
+      Log.error(error)
+      throw AdapterError.connlibConnectError(String(describing: error))
```
I'd consider logging and re-throwing the error an anti-pattern that probably leads to duplicate logs. I think just re-throwing is good enough here.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/AdapterUniFfi.swift:
```diff
@@ -0,0 +1,569 @@
+            Log.log("Event received: \(String(describing: event))")
+            await handleEvent(event)
+          }
+          // If no event, continue polling immediately
```
I think no event means that the receiver is closed so we should stop the loop. That is what we do on Android: https://github.com/firezone/firezone/blob/7222167b1396ca464ef8a03e783071f69432f0a7/kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt#L548-L551

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/AdapterUniFfi.swift:
```diff
@@ -0,0 +1,569 @@
+      }
+      provider.onUpdateResources(resourceList: resources)
+
+    case .disconnected(let error):
```
We should stop the event loop once we receive disconnected.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/AdapterUniFfi.swift:
```diff
@@ -0,0 +1,569 @@
+  #endif
+
+  #if os(iOS)
+    private func getSystemResolversViaBindResolvers() -> [String] {
```
None of this stuff is new right? It would be nice if we could somehow get `git` to recognise that this code hasn't changed. Maybe we can reorder this a bit so it mimics the old `Adapter` file?

klochowicz replied:
I've renamed it back to Adapter.swift (so it at least started recognising it) and currently updating it to at least recognize most of it. won't be perfect, but at least bulk should be noticed

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -361,3 +405,95 @@ extension TunnelConfiguration {
+
+// MARK: - Event Handlers from AdapterUniFfi
+
+extension PacketTunnelProvider {
```
Why is this an extension? Can't we just add the methods to the original class?

klochowicz replied:
it was just logically different, handlers of events; even put a different MARK for these. 
happy to fold them if you dislike this grouping.

thomaseizinger replied:
Don't mind much, was just wondering why :)

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -361,3 +405,95 @@ extension TunnelConfiguration {
+      // Now trigger the TUN device setup
+      Log.log("Triggering TUN device setup...")
+      adapter?.setupTunDevice()
+    }
```
I am confused as to why we need this. We should be able to set up the TUN device as soon as `connect` finishes.

klochowicz replied:
indeed. this was me overcomplicating stuff when I couldn't get it to work initially....

thomaseizinger reviewed (commented):
Great work! Looking forward to this landing. I've left a few comments with questions / suggestions.

jamilbk commented on swift/apple/FirezoneKit/Package.resolved:
Hmm I think this is redundant with the workspace-wide Package.resolved?

klochowicz replied:
I think this appeared when I added sentry dependency to FirezoneKit. 

this was a solution to missing dependency errors I would get after cleaning the build, to which I had to do "File -> Reset Package Cache" quite often.

perhaps instead I should have declared the sentry  dependency project wide too?

FWIW: I've learned very recently that dependabot looks into `FirezoneKit` directory only to update swift packages

jamilbk replied:
Ah! Yes, I've hit this too. We have had issues getting Dependabot to find/recognize Swift dependencies.

For Apple platforms specifically we've avoided third-party libs where possible since the stdlib has essentially everything you need, and the Swift package ecosystem / manager is not great, and dependabot was having issues.

If you want to move Sentry to a project-wide Package.swift instead of inside FirezoneKit, and that alleviates both issues, then 👍

klochowicz replied:
I've looked into this a bit deeper, and given sentry is only the dependency of FirezoneKit, IMHO it makes more sense to contain it in there and get rid of project-wide dependency (thus removing the workspace `Package.resolved`). 

This seems to still reliably avoid the package cache issues and keeps the project a bit "simpler" by encapsulating sentry lower in the hierarchy, where it is actually used. 

I've kept it here in this PR, but happy to move it to a separate one for dedicated discussion!

thomaseizinger replied:
> I've kept it here in this PR, but happy to move it to a separate one for dedicated discussion!

What I'd suggest is, keep it in here for now so you can iterate better but post another PR in parallel to land it in main. If we need to bisect something, it is always easier to have focused commits in main. Personally, I also like shipping small things to main as part of working on something bigger :)

klochowicz replied:
I agree with this sentiment, made the PR last week with extracted minor refactorings:  #10461

jamilbk commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -93,23 +117,30 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
+
+      try adapter.start(
+        apiUrl: apiURL,
+        token: token.description,
```
Curious why `.description` was appended here? Did the type change from String, or I guess that's due to the new UniFFI type?

klochowicz replied:
`Token` was always the underlying type, but FFI needs String eventually - I can still pass `Token` here and move the conversion at the `Session.newIos()`, it's a bit more clear then.

jamilbk commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -93,23 +117,30 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
+        Log.log("Initial internet resource enabled set to: \(internetResourceEnabled)")
+      } catch {
+        Log.error(error)
+      }
```
Hmmm perhaps we want to let the catch block just below this handle any possible failures here? I.e. be noisy if we fail to setInternetResourceEnabled?

klochowicz replied:
restructed this, no need for a catch block anymore.

jamilbk commented on swift/apple/Makefile:
With the scripts added in this PR, we now have the following places where we build macOS / iOS products:


- scripts/build/ios-appstore.sh
- scripts/build/macos-appstore.sh
- scripts/build/macos-standalone.sh
- scripts/build/uniffi-ios.sh
- scripts/build/uniffi-macos.sh
- swift/apple/Makefile
- swift/apple/xcode-build-connlib.sh
- .github/workflows/_swift.yml

Before, to build/run a debug app, you simply hit the "Build & Run" button from Xcode and this would kick off the `build-rust.sh` as a Build Phase prior to building the Swift sources.

The other scripts in `scripts/` are used to do release builds from CI and local macs that have the release certs/profiles on them. Release builds are an entirely different beast and involve a quite complex process of notarization, DMG compilation, pushing to github release, etc.

It would probably be worth pairing up next week to debug why your sysex is not automatically loading for your debug builds since this seems to be the reason some of this extra complexity is added here if I'm understanding things correctly.

Debugging signing / notarization issues is extremely obtuse when it comes to System Extensions since they run as root and have very strict security requirements, which Apple's documentation does a poor job of explaining.

Maybe a good thing to check off in this PR is that running `scripts/build/macos-standalone.sh` actually gets you a working release build. If we get there, we can be reasonably assured the build pipeline is working reliably.

thomaseizinger replied:
> the reason some of this extra complexity is added here if I'm understanding things correctly.

As far as I understand, the new `uniffi-*.sh` scripts are necessary to generate the UniFFI glue code. For Android, we do this as a gradle task and hence it is not as visible.

jamilbk replied:
I see - perhaps these could be consolidated into a single script and called as a Build Phase like the previous build-rust.sh script instead?

jamilbk replied:
The nice thing about the build phase is that it gets triggered both from Xcode UI builds and `xcodebuild` CLI builds. The Xcode project file configures that which is a little obtuse to read in a diff but otherwise maintainable as plaintext.

klochowicz replied:
apparently there's a project to keep it readable in YAML/json and only generate from that for XCode: https://github.com/yonaskolb/XcodeGen

I've reduced the duplication; most of it was leftovers from the xcframework process that was a total overkill. in the end, deleting dylibs was easiest way to keep a single `client-ffi` crate. 

 we now have only one script (just like before). If you don't mind, I'll keep it in the `swift/apple/build-rust.sh` location to make it easier to reason about (without that breadcrumb it was pretty hard to figure everything out).

and it's invoked from XCode both via CLI and from the UI; tested with both iOS and macOS.

jamilbk commented on rust/apple-client-ffi/build-rust.sh:
```diff
@@ -1,117 +0,0 @@
-
-cmd=${1:-""}
-
-# Sanitize the environment to prevent Xcode's shenanigans from leaking
```
So these were added as the result of a long debugging session that uncovered some kind of linking issue between iOS libs and rustc (might have been `ring` IIRC?).

Would definitely love to remove these but just adding this note here as a reminder to test both debug/release builds on iOS. I'll make sure to do this as well.

klochowicz replied:
Thank you for the background on this. I suspect the one that gave the most value was the sanitization of $PATH, which I kept too. 

I'll bring in back as it was here; and see whether we can get rid of it in a small targetted follow-up PR.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/AdapterUniFfi.swift:
```diff
@@ -0,0 +1,569 @@
+
+    // Create the session
+    do {
+      session = try Session.newIos(
```
A general note on the design here:

What I did for Android was spawn a dedicated event-loop task that "owns" the `Session` instance. When we bail out of that loop for whatever reason, the OS will garbage-collect the `Session`, which triggers `Drop` on the Rust side. This is quite important so we don't have memory leaks across sessions.

I think it would be useful to do the same thing for Apple:

- Spawn a task that holds the session
- Define one channel that sends commands in
- Create another channel that sends the events out

Does Swift have channels in its stdlib?

klochowicz replied:
yes, I believe it has channels. I'll look into it.

thomaseizinger replied:
I am also open to other designs, we just need to make sure we always correctly free all the memory (and make sure it is memory safe).

klochowicz commented on swift/apple/FirezoneNetworkExtension/AdapterUniFfi.swift:
```diff
@@ -0,0 +1,569 @@
+      guard addrBytes.count == 4 else { return false }
+      let firstByte = addrBytes[0]
+      return firstByte == 100  // 100.x.x.x (RFC 6598)
+        || firstByte == 127  // 127.x.x.x (loopback)
```
restored old sentinel range detection.

thomaseizinger commented on rust/client-ffi/.cargo/config.toml:
```diff
@@ -0,0 +1,12 @@
+# This must match the Xcode project setting (12.4)
+MACOSX_DEPLOYMENT_TARGET = "12.4"
+
+[build]
+# Use incremental compilation for faster rebuilds
+incremental = true
```
I am surprised this is not the default?

klochowicz replied:
it's default only in debug builds, not release (in case you want to do profiling)

thomaseizinger commented on rust/client-ffi/uniffi.toml:
```diff
@@ -0,0 +1,16 @@
+
+[bindings.kotlin]
+# Kotlin configuration (already working)
+async_runtime = "tokio"
```
Nit: Missing newline.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -93,22 +111,24 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
-
-      // Tell the system the tunnel is up, moving the tunnel manager status to
-      // `connected`.
+      // Report success immediately - TUN setup will happen when we receive the first interface config
```
This bit I don't understand. We should not need to do this asynchronously. The TUN file descriptor should already be active, same as before so we should just be able to search for it and set it as part of `start`?

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -361,3 +390,78 @@ extension TunnelConfiguration {
+    // The adapter now handles storing and hash comparison for the resource list
+    // The resourceList is already stored in the adapter as resourceListJSON
+    // and will be used for hash comparison in getResourcesIfVersionDifferentFrom
+    Log.log("Resources updated, stored in adapter for hash comparison")
```
This log seems a bit dodgy. Why is this here and not where the actual storing of the hash happens?

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionManager.swift:
```diff
@@ -0,0 +1,250 @@
+      )
+    } catch {
+      Log.error(error)
+      throw AdapterError.connlibConnectError(String(describing: error))
```
Log and rethrow is an anti-pattern IMO.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionManager.swift:
```diff
@@ -0,0 +1,250 @@
+      throw AdapterError.connlibConnectError(String(describing: error))
+    }
+
+    self.session = session
```
I am not sure I understand why are saving the session here. The point of the event-loop is to own the session variable, right? Why do we need another stateful object (the `SessionManager`) when all we want to do is spawn an async function (the event-loop) as a `Task`?

It seems to me like this `SessionManager` could be folded into the adapter where the other ends of the channel are owned by the adapter and we have a free-function that acts as the event-loop.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionManager.swift:
```diff
@@ -0,0 +1,250 @@
+        guard let self = self else { return }
+
+        var eventCount = 0
+        var pollAttempts = 0
```
Is the purpose of this just logging? If so, I think we can remove it to keep things simpler.

klochowicz replied:
yup, just logging; wanted to make sure I'm actually polling appropriately. removed :)

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionManager.swift:
```diff
@@ -0,0 +1,250 @@
+      }
+
+      // Wait for first task to complete, then cancel all
+      _ = await group.next()
```
Nice!

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionManager.swift:
```diff
@@ -0,0 +1,250 @@
+    // Do NOT call disconnect() explicitly - let Drop handle everything
+    Log.log("SessionManager: Event loop finished, cleaning up session")
+    self.session = nil
+    Log.log("SessionManager: Session cleaned up")
```
This seems redundant and is the kind of stuff I'd like to avoid. If we "move" the session into the event-loop, breaking from both loops should reduce the reference count to zero and free all the memory, meaning we don't need to actually assign something to `nil` to free the last reference.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionManager.swift:
```diff
@@ -0,0 +1,250 @@
+
+    // Cancel event loop task
+    eventLoopTask?.cancel()
+    eventLoopTask = nil
```
I am not super familiar with how this works in Swift but I'd expect that we don't need to cancel this because the event-loop should exit by itself after receiving `disconnect`.

I am not sure about the pointer to the task. Do we need to keep it around at all or can we perhaps get rid of the field entirely? In Kotlin, I think we are just detaching the task and let it run in the background.

klochowicz replied:
you're right; we shouldn't need this! 

pointer to the task for only for logging; it will happily run itself detached. 

reworked to address feedback

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionManager.swift:
```diff
@@ -0,0 +1,250 @@
+    eventLoopTask = nil
+
+    // Close command stream
+    commandContinuation.finish()
```
Coming from Rust, I think this should be called `sender`.

klochowicz replied:
I tried to preserve idiomatic naming for Swift; this is how it's called. it behaves a bit differently than a Sender; notably, dropping Continuation doesn't close the channel :(

klochowicz replied:
based on OOB discussion, I'll wrap in something that resembles a Rust channel.

klochowicz replied:
I ended up writing a `Channel` "adapter" that works more like Rust with its RAII

thomaseizinger commented on swift/apple/Makefile:
```diff
@@ -137,3 +193,37 @@ tail-tunnel-log:
+	@echo "  make build-macos                  # Build only, don't install"
+	@echo "  make install                      # Install previously built app"
+	@echo "  make show-client-log              # View latest app log file"
+	@echo "  make show-ext-log                 # View network extension logs"
```
This is prone to get outdated I reckon but happy to keep it in here if you want it.

klochowicz replied:
ideally, we'd move to `just` or `mise` at some point, so this wouldn't be needed at all, as it's generated on-the-fly. 

if you don't mind, I'll keep it for a bit as it can be helpful when we do the change.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -39,34 +39,26 @@ enum AdapterError: Error {
-
-  /// Network settings
-  private var networkSettings: NetworkSettings?
+/// Adapter using UniFFI polling-based event handling
```
Redundant comment IMO.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -144,40 +136,65 @@ class Adapter {
+        )
+      } catch {
+        Log.error(error)
+        throw error
```
If you remove the logging here, you don't even need the catch block.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -144,40 +136,65 @@ class Adapter {
+    }
+  }
 
-  init(
```
Nit: Can we keep the original `init` / `start` split or does that not work? Would probably reduce the diff a bit further.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionManager.swift:
```diff
@@ -0,0 +1,250 @@
+  init(eventHandler: @escaping (Event) async -> Void) {
+    self.eventHandler = eventHandler
+    (commandStream, commandContinuation) = AsyncStream.makeStream()
+  }
```
Instead of a callback, I think it would be nicer to also have an `AsyncStream` here.

thomaseizinger reviewed (commented):
Great progress! I've left a few more comments that should hopefully simplify the design a bit.

The main things are:

- I don't think we need the explicit, `searchForTunFD` thing. We should just be able to do that inline as part of startup the way we did before
- The `SessionManager` seems like an unnecessary indirection
- Events coming from connlib don't yet use a channel but still a callback

klochowicz committed:
fix(apple): Add missing is_internet_resource_active parameter and simplify code

- Add is_internet_resource_active parameter to new_apple constructor
- Remove unused internetResource property
- Remove unused resources() helper method
- Simplify resource tracking to only store JSON for UI consumption

klochowicz committed:
fix(apple): Fix Profile build

klochowicz committed:
ci: Don't mention apple-client-ffi

The crate doesn't exist anymore

klochowicz committed:
chore: Run cargo-sort

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -139,85 +146,102 @@ class Adapter {
-    guard session == nil else {
-      throw AdapterError.invalidSession(session)
+    do {
+      try session.setTunFromSearch()
```
Can we just call `Tun::new` inside `Session.newApple` on the Rust side?

Eventually (when we build a headless Apple client), we might want to change that but for this PR, it seems simpler to stick to what we are currently doing because it means we don't have to introduce this `setTunFromSearch` function.

klochowicz replied:
yes, we could do it on the Rust side. I've kept the function with the retry logic there though, I just call it directly now. 

the main footgun around here was the fact that I needed a tokio runtime for `Tun::new()` - and it was pretty nasty, because the network extension crashing due to that is not presented well in our main app (looks we're just being signed out).

I've added explicit passing of the runtime for both Android and Swift to self-document this requirement (we need it only to spawn some otel tasks)

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -139,85 +146,102 @@ class Adapter {
   private let logFilter: String
-  private let connlibLogFolderPath: String
 
-  init(
```
Nit: A few more re-orderings and renamings of variable would reduce the diff here!

We are very close :)

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -35,34 +35,41 @@ enum AdapterError: Error {
   /// Network routes monitor.
   private var networkMonitor: NWPathMonitor?
 
-  /// Used to avoid path update callback cycles on iOS
```
I am surprised to see this gone!

thomaseizinger replied:
That is probably why the iOS build is failing?

klochowicz replied:
it's failing due to not finished moving parts (conditional compilation 🙈 ) . should be fixed in a jiffy!

klochowicz replied:
> I am surprised to see this gone!

I believe it should have been removed with these changes: https://github.com/firezone/firezone/pull/10022/files#diff-a84e8f62a17ac67f781019e6ac0456567fd18ffa7c13b3248609d78debb6480eL342

we now prevent cycles on iOS this way https://github.com/firezone/firezone/commit/290b0c298f7d3345d3376f6b20c362825298019f ?

thomaseizinger replied:
Ah interesting! Good catch.

jamilbk replied:
Yes! This is correct - good catch.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -17,6 +17,8 @@ enum PacketTunnelProviderError: Error {
 
 class PacketTunnelProvider: NEPacketTunnelProvider {
   private var adapter: Adapter?
+  private var hasReceivedFirstConfig = false
```
Do we still need this?

klochowicz replied:
thanks for spotting, why swift doesn't spot unused variables is just beyond me 🤷‍♂️ 

maybe it's fixed in new Swift or some other tooling.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -17,6 +17,8 @@ enum PacketTunnelProviderError: Error {
 class PacketTunnelProvider: NEPacketTunnelProvider {
   private var adapter: Adapter?
+  private var hasReceivedFirstConfig = false
+  private var networkSettings: NetworkSettings?
```
It seems the `networkSettings` have moved to this file. Is there a reason why it couldn't be in `Adapter` anymore?

klochowicz replied:
moved it back

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -35,6 +37,14 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
 
     super.init()
 
+    // Log version information immediately on startup
```
I don't mind adding these logs but I also wouldn't be opposed to leaving them out of this PR.

klochowicz replied:
I'll keep them here if you don't mind

thomaseizinger reviewed (commented):
Thank you for all the effort in reducing the diff of this PR. It is much easier to review to now.

klochowicz committed:
chore: Inline some handlers

This brings back networkSettings into Adapter, reducing diff size.

klochowicz committed:
chore: Don't go via FFI to set up the tunnel

Can be called directly

klochowicz committed:
chore: Log successful update of the network extension

klochowicz committed:
refactor(client-ffi): Pass runtime handle explicitly to Tun constructors

Previously, `Tun::new()` and `Tun::from_fd()` relied on ambient Tokio
runtime context via `.enter()` guards. This was a footgun - calling
these functions without an active runtime would panic at runtime with
an obscure error.

Now we pass `&tokio::runtime::Handle` explicitly to all Tun
constructors. This makes the runtime dependency clear in the function
signature and prevents accidental misuse.

Benefits:
- Compile-time enforcement of runtime requirement
- Self-documenting API
- No hidden ambient state or panics
- More idiomatic Rust (explicit dependencies)

klochowicz committed:
chore: Remove unused variable

klochowicz committed:
chore: Reduce diff noise

klochowicz commented on rust/client-ffi/src/platform/apple/tun.rs:
```diff
@@ -20,22 +20,38 @@ pub struct Tun {
 
 impl Tun {
-    pub fn new() -> io::Result<Self> {
+    pub fn new(runtime: &tokio::runtime::Handle) -> io::Result<Self> {
```
@thomaseizinger I've opted in for passing this runtime explicitly to make it clear that we need runtime here (both for Android and Swift) as I've been hit by this footgun - especially when NetworkExtension crashes silently in the background 😢

thomaseizinger replied:
Is passsing the runtime handle enough? I thought we'd also need to `enter` it?

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -258,119 +277,102 @@ class Adapter {
+      if let routesData = ipv4Routes.data(using: .utf8),
+        let cidrs = try? JSONDecoder().decode([NetworkSettings.Cidr].self, from: routesData)
+      {
+        networkSettings.routes4 = cidrs.compactMap({ $0.asNEIPv4Route })
```
I think we should avoid making partial updates to the `networkSettings` here. Instead, I think we should try and decode everything first into local variables and then only if it all worked, try and apply it.

thomaseizinger replied:
We can throw an exception here as long as we wrap the call-site in a `try catch`.

klochowicz replied:
makes sense! will restore the previous behaviour

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -139,85 +149,95 @@ class Adapter {
-      let msg = (error as? RustString)?.toString() ?? "Unknown error"
-      throw AdapterError.connlibConnectError(msg)
+      for await event in eventReceiver.stream {
+        await self?.handleEvent(event)
```
Can `handleEvent` fail? If yes, we should wrap this in a try catch and log an error but continue operating.

thomaseizinger replied:
Given that `self` is a weak-reference here, I think we should use a `guard` that breaks out of the loop if `self` ends up being nil. The guard needs to be inside the loop though to make sure we check it on each iteration.

klochowicz replied:
> Can `handleEvent` fail? If yes, we should wrap this in a try catch and log an error but continue operating.

yes, it can fail, only if the JSON decoding fails - the other case is wrapped in try/catch block.  I guess if we address #9548 it will sort that part out? 
Swift compiler actually would have caught this if it actually could throw an error and we're not dealing with it. 

> Given that self is a weak-reference here, I think we should use a guard that breaks out of the loop if self ends up being nil. The guard needs to be inside the loop though to make sure we check it on each iteration.

good catch.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -139,85 +149,95 @@ class Adapter {
+    self.commandSender = commandSender
+
+    let (eventSender, eventReceiver): (Sender<Event>, Receiver<Event>) = Channel.create()
+    self.eventReceiver = eventReceiver
```
Do we need to store the event receiver in `self`? I think we could just pass it on as a local variable to the task that processes it. That should give us one less thing to clean up.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -139,85 +149,95 @@ class Adapter {
+  }
+
+  // Could happen abruptly if the process is killed.
+  deinit {
```
I am not entirely sure how Swift's ownership model works but we should make sure to cancel all tasks and free all resources here.

klochowicz replied:
yeah, it seems to be suggested to keep the reference to tasks and cancel them explicitly here; I'll do that.

thomaseizinger reviewed (approved):
Pushed two commits to reduce the diff a bit more. I think this is ready. A few final questions regarding resource clean-up!

Great work, thank you for all the effort and bearing with all my comments :)

jamilbk reviewed (commented):
Took one more glance over it and I don't see anything on the Swift side that jumps out as an issue. Will run it through the paces once a build hits TestFlight.

klochowicz committed:
chore: Bring back old way of network setting update

klochowicz committed:
chore: Don't store the event receiver locally

It can be passed directly to the task, and used there.
One less thing to worry about cleaning up.

klochowicz committed:
refactor(apple): improve Adapter lifecycle management and error handling

- Store Task handles for explicit cancellation in deinit
- Remove eventReceiver from stored properties (now local variable)
- Add guard let self check inside event loop to exit when Adapter is deallocated

klochowicz committed:
chore: Move deinit after init

klochowicz committed:
chore: Move some things around

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionEventLoop.swift:
```diff
@@ -0,0 +1,91 @@
+        } catch {
+          Log.error(error)
+          Log.log("SessionEventLoop: Error in event polling, exiting event loop")
+          break
```
I don't think this is what we want. I think we should continue on error.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionEventLoop.swift:
```diff
@@ -0,0 +1,91 @@
+            // No event returned - session has ended
+            Log.log("SessionEventLoop: Event stream ended, exiting event loop")
+            break
+          }
```
Nit: I think this would be cleaner with a `guard`.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -258,119 +296,91 @@ class Adapter {
-        networkSettings ?? NetworkSettings(packetTunnelProvider: packetTunnelProvider)
-
-      guard let data4 = routeListv4.data(using: .utf8),
-        let data6 = routeListv6.data(using: .utf8),
+      // Decode all data into local variables first to ensure all parsing succeeds before applying
+      guard let dnsData = dns.data(using: .utf8),
+        let dnsAddresses = try? JSONDecoder().decode([String].self, from: dnsData),
+        let data4 = ipv4Routes.data(using: .utf8),
+        let data6 = ipv6Routes.data(using: .utf8),
         let decoded4 = try? JSONDecoder().decode([NetworkSettings.Cidr].self, from: data4),
         let decoded6 = try? JSONDecoder().decode([NetworkSettings.Cidr].self, from: data6)
       else {
-        fatalError("Could not decode route list from connlib")
+        fatalError("Could not decode network configuration from connlib")
```
What is a fatalError in Swift? Is that a panic / exception?

thomaseizinger replied:
This was already there before so doesn't need to change now. But I think it would be good to handle this gracefully here but log a fat error (that we'll see in Sentry).

klochowicz replied:
> What is a fatalError in Swift? Is that a panic / exception?

it's a panic / unrecoverable error.

Happy to deal with it in a follow-up.

thomaseizinger replied:
Yeah I think we should handle this more gracefully.

thomaseizinger commented on swift/apple/build-rust.sh:
I think this can be tidied up in a follow-up too. There is a fair amount of (IMO) unnecessary stuff in here (all the echo output, testing for generated files, comments).

klochowicz committed:
chore: Address review feedback

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/SessionEventLoop.swift:
```diff
@@ -0,0 +1,91 @@
+      for await command in commandReceiver.stream {
+        await handleCommand(command, session: session)
+
+        // Exit loop if disconnect command
+        if case .disconnect = command {
+          Log.log("SessionEventLoop: Disconnect command received, exiting command loop")
+          break
+        }
```
Technically, I don't think we need this. When we send `disconnect`, connlib's event-loop will exit, closing the stream which in turn causes the event-polling task to exit.

In fact, I think we should not `break` here and instead keep everything running until connlib has shut down (which we learn about from `session.nextEvent` returning `nil`.

thomaseizinger replied:
The worry I have here is that if we break here early, the entire task gets cancelled, dropping the last reference to `Session` which might abort connlib's graceful shutdown.

thomaseizinger reviewed (approved):
I think this is good to go in. It has received multiple reviews and there is probably not a whole lot more that can be spotted in the PR review itself.

What we should do is, once this is merged, do another pass over the entire Swift codebase concerning everything around `Adapter` with a more holistic view onto things.

My guess would be that we will spot more things then that we want to slightly change / tweak / fix.
